nvflare/app_common/abstract/learner2.py

```diff
@@ -0,0 +1,149 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import abstractmethod
+from typing import Union
+
+from nvflare.apis.dxo import DXO
+from nvflare.apis.fl_component import FLComponent
+from nvflare.apis.fl_context import FLContext
+
+
+class Learner2(FLComponent):
+
+    STATE = None
+
+    def __init__(self):
+        super().__init__()
+        self.engine = None
+        self.fl_ctx = None
+        self.workspace = None
+        self.shareable = None
+        self.args = None
+        self.site_name = None
+        self.job_id = None
+        self.app_root = None
+        self.abort_signal = None
+        self.current_round = 0
+        self.total_rounds = 0
+
+    def is_aborted(self) -> bool:
+        return self.abort_signal and self.abort_signal.triggered
+
+    def get_shareable_header(self, key: str, default=None):
+        if not self.shareable:
+            return default
+        return self.shareable.get_header(key, default)
+
+    def get_context_prop(self, key: str, default=None):
+        if not self.fl_ctx:
+            return default
+        assert isinstance(self.fl_ctx, FLContext)
+        return self.fl_ctx.get_prop(key, default)
+
+    def debug(self, msg: str):
+        self.log_debug(self.fl_ctx, msg)
+
+    def info(self, msg: str):
+        self.log_info(self.fl_ctx, msg)
+
+    def error(self, msg: str):
+        self.log_error(self.fl_ctx, msg)
+
+    def warning(self, msg: str):
+        self.log_warning(self.fl_ctx, msg)
+
+    def exception(self, msg: str):
+        self.log_exception(self.fl_ctx, msg)
+
+    def critical(self, msg: str):
+        self.log_critical(self.fl_ctx, msg)
+
+    def panic(self, reason: str):
+        self.system_panic(reason, self.fl_ctx)
+
+    def initialize(self, parts: dict):
```

Comment:
Maybe we could hide it in "self" and provide a "get_component" method. If the learner needs a component (e.g. log_writer), it can get it with the get_component method.

Reply:
Yes. It would still be nice to have a way to get the components.

Reply:
Changed as suggested.